README.md
# Overlay.js

Basic overlay module, built for easy extension through ES6 modules.

## How to use

### Javascript

```javascript
import Overlay from 'path/to/overlay';

var o = new Overlay(el, options);
o.render('<p>Hi</p>');
o.show();
setTimeout(function() {
  o.hide();
}, 2000);
```

You can import other extensions as well and use them as you wish.

### HTML

The most basic version is below. You can customise anything as long as you either slap the right classes on it or change the selectors in the JS options.

```html
<div class="mod-overlay js-overlay is-hidden">
  <div class="overlay__content js-overlay-content"></div>
</div>
```

### SCSS

There's a tiny bit of CSS that you might want to include so it works smoothly. The main part is the magic with the `transition` on the `z-index`. This allows for smooth CSS transitions, a switch on `display` does not.

```scss
.mod-overlay {
  // your styles here
  
  $transition-duration: 220ms;
  transition: opacity $transition-duration ease-in-out, z-index 16ms linear $transition-duration;
  z-index: -1; // Hide the overlay below everything

  &.is-hidden {
    opacity: 0;
  }

  &.is-shown {
    opacity: 1;
    z-index: 800;
    transition: opacity $transition-duration ease-in-out, z-index 16ms linear;
  }
}
```

## Installation

`$ bower install overlay-js --save`


HISTORY.md
# Changelog

## 1.0.5

- When providing a `<link rel="up" href="/foo/bar">` tag in the `<head>`, the overlay will fetch that page's contents and append it to the body

## 1.0.4

- `overlay:show` and `overlay:hide` event

## 1.0.3

- History routing with changing document title if included in the AJAX response

## 1.0.2
## 1.0.1

- HTML5 History routing

## 1.0.0

- **Complete rewrite** using ES6 modules and a new approach

---

## 0.2.3

- Fix hide_callback, clear_content was executed too early


## 0.2.2

- Fix content_key reset in the hide function (was executed too early)


## 0.2.1

- Fix hide_callback on browsers without transition support


## 0.2.0

- Removed the use of transitionend and used setTimeout instead.
- Added a setting to adjust the setTimeout duration
- Improved the content_key usage
- Added a setting to set the default content key
- Added a pre-hide callback


## 0.1.5

- Show callback isn't always triggered, transitionend listener is troublesome. Hotfixed by setting timeout to 200ms
- Added background close by default


## 0.1.4

- README fix


## 0.1.3

- Better explanation of how to use the plugin
- Show callback event is now `overlay.show.CONTENT_KEY` instead of `overlay.open.CONTENT_KEY`
- Hide callback event now also works with the content key (i.e. the same as the show callback)


## 0.1.2

- Also remove the background element when the
  destroy function is executed


## 0.1.1

- Also render templates for predefined elements
  -> And add option to disable that


## 0.1.0

`INITIAL_VERSION`


gulpfile.js
var gulp = require('gulp');
var sourcemaps = require('gulp-sourcemaps');
var source = require('vinyl-source-stream');
var buffer = require('vinyl-buffer');
var browserify = require('browserify');
var watchify = require('watchify');
var babel = require('babelify');

function compile(watch) {
  var bundler = watchify(browserify('./demo/demo.js', { debug: true }).transform(babel));

  function rebundle() {
    bundler.bundle()
      .on('error', function(err) { console.error(err); this.emit('end'); })
      .pipe(source('build.js'))
      .pipe(buffer())
      .pipe(sourcemaps.init({ loadMaps: true }))
      .pipe(sourcemaps.write('./'))
      .pipe(gulp.dest('./demo'));
  }

  if (watch) {
    bundler.on('update', function() {
      console.log('-> bundling...');
      rebundle();
    });
  }

  rebundle();
}

function watch() {
  return compile(true);
};

gulp.task('build', function() { return compile(); });
gulp.task('watch', function() { return watch(); });

gulp.task('default', ['watch']);


src/router.js
export default class Router {
  constructor (handlers) {
    this.isSupported = (!!history && !!history.pushState);
    this.root = {};

    this.handlers = {
      pop: handlers.pop || (function () {})
    };

    this.init().bind();
  }

  init () {
    this.root = {
      url: window.location.pathname,
      title: document.title
    };

    this.replace(this.root.url, { title: this.root.title });
    return this;
  }

  bind () {
    window.addEventListener('popstate', e => {
      console.log('ja dus');
      this.handlers.pop(e.state);
    });

    return this;
  }

  push (url, state = {}) {
    if (this.isSupported) {
      state.url = url;
      state.title = state.title || '';

      if (state.title.length > 0) {
        document.title = state.title;
      }

      window.history.pushState(state, state.title, url);
    }
  }

  replace (url, state = {}) {
    if (this.isSupported) {
      state.url = url;
      state.title = state.title || '';

      if (state.title.length > 0) {
        document.title = state.title;
      }

      window.history.replaceState(state, state.title, url);
    }
  }

  pushRoot () {
    this.push(this.root.url, { url: this.root.url, title: this.root.title });
  }

  setRoot (url, title = '') {
    this.root.url = url;
    this.root.title = title;
  }
}

overlay-es6.js
export default class Overlay {

  // @constructor
  // @param {HTMLElement} el
  // @param {Object}      options
  constructor (el = document.getElementsByClassName('js-overlay')[0], options = {}) {
    this.el = el;

    // @todo - Merge with options
    this.options = {
      selectors: {
        close: '.js-overlay-close',
        content: '.js-overlay-content',
        show: '.js-overlay-show',
        toggle: '.js-overlay-toggle'
      },
      states: {
        hidden: 'is-hidden',
        shown: 'is-shown'
      }
    };

    this.init().bind();
  }

  // Initial setup
  init () {
    return this;
  }

  // Bind event handlers to certain DOM elements, specified through
  bind () {
    this.bindDelegate('click', this.options.selectors.show, e => {
      e.preventDefault();
      e.stopPropagation();
      this.show();
    });

    this.bindDelegate('click', this.options.selectors.hide, e => {
      e.preventDefault();
      e.stopPropagation();
      this.hide();
    });

    this.bindDelegate('click', this.options.selectors.toggle, e => {
      e.preventDefault();
      e.stopPropagation();
      this.toggle();
    });

    return this;
  }

  // Helper method to support clicks on elements that don't exist yet
  // @param {String}    evt       Event to listen for
  // @param {String}    selector  Element selector that triggers the event
  // @param {Function}  handler   Callback method
  bindDelegate (evt, selector, handler) {
    let matches = document.body.matches || document.body.webkitMatchesSelector || document.body.msMatchesSelector;

    // Listen to all events of type evt
    document.body.addEventListener(evt, e => {
      // If e.target matches the specified selector or is a child element
      if (!!matches.call(e.target, `${selector}, ${selector} *`)) {
        // Execute handler, pass on the Event object
        handler(e);
      }
    }, false);
  }

  // Show the overlay
  show () {
    let regex = new RegExp("\\s*" + this.options.states.hidden + "\\s*", 'gi'); // Template string not working because of reasons
    this.el.className = this.el.className.replace(regex, '') + ' ' + this.options.states.shown;
  }

  // Hide the overlay
  hide () {
    let regex = new RegExp("\\s*" + this.options.states.shown + "\\s*", 'gi'); // Template string not working because of reasons
    this.el.className = this.el.className.replace(regex, '') + ' ' + this.options.states.hidden;
  }

  // Toggle the overlay
  toggle () {
    if (this.el.className.indexOf(this.options.states.shown) === -1) {
      this.show();
    } else {
      this.hide();
    }
  }

  // Render HTML inside the content selector
  render (html) {
    this.el.querySelector(this.options.selectors.content).innerHTML = html;
  }
}


src/overlay.js
export default class Overlay {

  // @constructor
  // @param {HTMLElement} el
  // @param {Object}      options
  constructor (el = document.getElementsByClassName('js-overlay')[0], options = { selectors: {}, states: {} }) {
    this.el = el;

    this.options = {
      selectors: {
        content: '.js-overlay-content',
        hide: '.js-overlay-hide',
        show: '.js-overlay-show',
        toggle: '.js-overlay-toggle'
      },
      states: {
        hidden: 'is-hidden',
        shown: 'is-shown'
      }
    };

    // Merge provided selectors
    for (let selector in options.selectors) {
      if (options.selectors.hasOwnProperty(selector)) {
        this.options.selectors[selector] = options.selectors[selector];
      }
    }

    // Merge provided states
    for (let state in options.states) {
      if (options.states.hasOwnProperty(state)) {
        this.options.states[state] = options.states[state];
      }
    }

    this.init().bind();
  }

  // Initial setup
  init () {
    return this;
  }

  // Bind event handlers to certain DOM elements, specified through
  bind () {
    document.body.addEventListener('keyup', e => {
      if (e.keyCode === 27) {
        this.hide();
      }
    });

    this.bindDelegate('click', this.options.selectors.show, e => {
      e.preventDefault();
      this.show();
    });

    this.bindDelegate('click', this.options.selectors.hide, e => {
      e.preventDefault();
      this.hide();
    });

    this.bindDelegate('click', this.options.selectors.toggle, e => {
      e.preventDefault();
      this.toggle();
    });

    return this;
  }

  // Helper method to support clicks on elements that don't exist yet
  // @param {String}    evt       Event to listen for
  // @param {String}    selector  Element selector that triggers the event
  // @param {Function}  handler   Callback method
  bindDelegate (evt, selector, handler) {
    let matches = document.body.matches || document.body.webkitMatchesSelector || document.body.msMatchesSelector;

    // Listen to all events of type evt
    document.body.addEventListener(evt, e => {
      // If e.target matches the specified selector or is a child element
      if (!!matches.call(e.target, `${selector}, ${selector} *`)) {
        // Find the actual element that matches the selector
        let matchingElement = e.target;
        while (!matches.call(matchingElement, selector)) {
          matchingElement = matchingElement.parentNode;
        }

        // Execute handler, pass on the Event object
        // and have the matching element as the second argument
        handler(e, matchingElement);
      }
    }, false);
  }

  // Show the overlay
  show () {
    if (this.el.className.indexOf(this.options.states.shown) !== -1) {
      return false;
    }

    let regex = new RegExp("\\s*" + this.options.states.hidden + "\\s*", 'gi'), // Template string not working because of reasons
        evt = new CustomEvent('overlay:show', { bubbles: true, cancelable: true }); // @fixme - IE9 support

    this.el.className = this.el.className.replace(regex, '') + ' ' + this.options.states.shown;

    document.body.dispatchEvent(evt);

    return true;
  }

  // Hide the overlay
  hide () {
    if (this.el.className.indexOf(this.options.states.hidden) !== -1) {
      return false;
    }

    let regex = new RegExp("\\s*" + this.options.states.shown + "\\s*", 'gi'), // Template string not working because of reasons
        evt = new CustomEvent('overlay:hide', { bubbles: true, cancelable: true }); // @fixme - IE9 support

    this.el.className = this.el.className.replace(regex, '') + ' ' + this.options.states.hidden;

    document.body.dispatchEvent(evt);

    return true;
  }

  // Toggle the overlay
  toggle () {
    if (this.el.className.indexOf(this.options.states.shown) === -1) {
      this.show();
    } else {
      this.hide();
    }
  }

  // Render HTML inside the content selector
  render (html) {
    this.el.querySelector(this.options.selectors.content).innerHTML = html;
  }
}


src/ajax-overlay.js
import Overlay from './overlay';
import Router from './router';

export default class AjaxOverlay extends Overlay {
  constructor (el, options = { selectors: {}, states: {} }) {
    if (!options.selectors.fetch) {
      options.selectors.fetch = '.js-overlay-fetch';
    }

    super(el, options);

    this.router = new Router({
      pop: (state) => { this.handleRoutePop(state) }
    });
  }

  init () {
    super.init();

    if (!!document.querySelector('link[rel="up"]')) {
      let root = document.querySelector('link[rel="up"]').getAttribute('href'),
          xhr = new XMLHttpRequest();

      xhr.open('GET', root, true);

      xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');

      xhr.onreadystatechange = () => {
        if (xhr.readyState !== 4 || xhr.status !== 200) {
          return false;
        }

        // If the <title> tag is present in the responseText
        // use that as the title in the browser / history
        let title = /<title>(.*)<\/title>/gi.exec(xhr.responseText),
            body = document.createElement('div');

        if (!!title && title.length >= 2) {
          title = title[1];
        } else {
          title = document.title;
        }

        this.router.setRoot(root, title);

        body.innerHTML = xhr.responseText;
        document.body.appendChild(body);
      }

      xhr.send();
    }

    return this;
  }

  bind () {
    super.bind();

    this.bindDelegate('click', this.options.selectors.fetch, (e, matchingElement) => {
      e.preventDefault();
      e.stopPropagation();

      let href = matchingElement.getAttribute('href');

      // Automatically load href attribute if it's on the trigger element
      if (!!href && href.length > 1) {
        this.fetch(href);
      }
    });

    return this;
  }

  // Fetch HTML from an URL and render it inside the overlay
  // @param {String}  url   Not supporting CORS properly, so try using relative paths altogether
  fetch (url) {
    let xhr = new XMLHttpRequest();

    xhr.open('GET', url, true);

    xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');

    xhr.onreadystatechange = () => {
      if (xhr.readyState !== 4 || xhr.status !== 200) {
        return false;
      }

      // If the <title> tag is present in the responseText
      // use that as the title in the browser / history
      let title = /<title>(.*)<\/title>/gi.exec(xhr.responseText);

      if (!!title && title.length >= 2) {
        title = title[1];
      } else {
        title = document.title;
      }

      // Expects HTML as responseText
      this.render(xhr.responseText);
      this.show(url, title);
    }

    xhr.send();
  }

  // Show overlay
  // @param {String}  url     URL to keep in the browser history
  // @param {String}  title   Title to associate with the URL in the browser history
  // @param {Bool}    push    Push to browser history?
  show (url = '', title = '', push = true) {
    if (super.show() && !!push) {
      this.router.push(url, { title: title, shown: true });
    }
  }

  // Hide overlay
  // @param {Bool}    push    Push to browser history?
  hide (push = true) {
    if (super.hide() && !!push) {
      this.router.pushRoot();
    }
  }

  // Handle browser popstate event
  // @param {Object} state  The state object associated with the popstate event
  handleRoutePop (state) {
    if (!!state.title && state.title.length > 0) {
      document.title = state.title;
    }

    if (!!state && !!state.shown) {
      this.show(state.url, state.title, false);
    } else {
      this.hide(false);
    }
  }
}


demo/build.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _srcAjaxOverlay = require('../src/ajax-overlay');

var _srcAjaxOverlay2 = _interopRequireDefault(_srcAjaxOverlay);

var overlay = new _srcAjaxOverlay2['default']();

document.body.addEventListener('overlay:show', function (e) {
  console.log('Overlay shown');
});

document.body.addEventListener('overlay:hide', function (e) {
  console.log('Overlay hidden');
});

if (!!document.querySelector('.js-overlay-toggle')) {
  document.querySelector('.js-overlay-toggle').addEventListener('click', function (e) {
    e.preventDefault();
    overlay.render('\n      <h2>This is a HTML string rendered inside the overlay</h2>\n      <p>It\'s not neccessary to render from JS, you can also just render static content and use the overlay module as a show / hide toggle.</p>\n      <p>Or you could use a template rendering engine like <a href="http://handlebarsjs.com/">Handlebars</a></p>\n    ');
  });
}

},{"../src/ajax-overlay":2}],2:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x6, _x7, _x8) { var _again = true; _function: while (_again) { desc = parent = getter = undefined; _again = false; var object = _x6,
    property = _x7,
    receiver = _x8; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x6 = parent; _x7 = property; _x8 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; }

var _overlay = require('./overlay');

var _overlay2 = _interopRequireDefault(_overlay);

var _router = require('./router');

var _router2 = _interopRequireDefault(_router);

var AjaxOverlay = (function (_Overlay) {
  function AjaxOverlay(el) {
    var _this2 = this;

    var options = arguments[1] === undefined ? { selectors: {}, states: {} } : arguments[1];

    _classCallCheck(this, AjaxOverlay);

    if (!options.selectors.fetch) {
      options.selectors.fetch = '.js-overlay-fetch';
    }

    _get(Object.getPrototypeOf(AjaxOverlay.prototype), 'constructor', this).call(this, el, options);

    this.router = new _router2['default']({
      pop: function pop(state) {
        _this2.handleRoutePop(state);
      }
    });
  }

  _inherits(AjaxOverlay, _Overlay);

  _createClass(AjaxOverlay, [{
    key: 'init',
    value: function init() {
      var _this3 = this;

      _get(Object.getPrototypeOf(AjaxOverlay.prototype), 'init', this).call(this);

      if (!!document.querySelector('link[rel="up"]')) {
        (function () {
          var root = document.querySelector('link[rel="up"]').getAttribute('href'),
              xhr = new XMLHttpRequest();

          xhr.open('GET', root, true);

          xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');

          xhr.onreadystatechange = function () {
            if (xhr.readyState !== 4 || xhr.status !== 200) {
              return false;
            }

            // If the <title> tag is present in the responseText
            // use that as the title in the browser / history
            var title = /<title>(.*)<\/title>/gi.exec(xhr.responseText),
                body = document.createElement('div');

            if (!!title && title.length >= 2) {
              title = title[1];
            } else {
              title = document.title;
            }

            _this3.router.setRoot(root, title);

            body.innerHTML = xhr.responseText;
            document.body.appendChild(body);
          };

          xhr.send();
        })();
      }

      return this;
    }
  }, {
    key: 'bind',
    value: function bind() {
      var _this4 = this;

      _get(Object.getPrototypeOf(AjaxOverlay.prototype), 'bind', this).call(this);

      this.bindDelegate('click', this.options.selectors.fetch, function (e, matchingElement) {
        e.preventDefault();
        e.stopPropagation();

        var href = matchingElement.getAttribute('href');

        // Automatically load href attribute if it's on the trigger element
        if (!!href && href.length > 1) {
          _this4.fetch(href);
        }
      });

      return this;
    }
  }, {
    key: 'fetch',

    // Fetch HTML from an URL and render it inside the overlay
    // @param {String}  url   Not supporting CORS properly, so try using relative paths altogether
    value: function fetch(url) {
      var _this5 = this;

      var xhr = new XMLHttpRequest();

      xhr.open('GET', url, true);

      xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');

      xhr.onreadystatechange = function () {
        if (xhr.readyState !== 4 || xhr.status !== 200) {
          return false;
        }

        // If the <title> tag is present in the responseText
        // use that as the title in the browser / history
        var title = /<title>(.*)<\/title>/gi.exec(xhr.responseText);

        if (!!title && title.length >= 2) {
          title = title[1];
        } else {
          title = document.title;
        }

        // Expects HTML as responseText
        _this5.render(xhr.responseText);
        _this5.show(url, title);
      };

      xhr.send();
    }
  }, {
    key: 'show',

    // Show overlay
    // @param {String}  url     URL to keep in the browser history
    // @param {String}  title   Title to associate with the URL in the browser history
    // @param {Bool}    push    Push to browser history?
    value: function show() {
      var url = arguments[0] === undefined ? '' : arguments[0];
      var title = arguments[1] === undefined ? '' : arguments[1];
      var push = arguments[2] === undefined ? true : arguments[2];

      if (_get(Object.getPrototypeOf(AjaxOverlay.prototype), 'show', this).call(this) && !!push) {
        this.router.push(url, { title: title, shown: true });
      }
    }
  }, {
    key: 'hide',

    // Hide overlay
    // @param {Bool}    push    Push to browser history?
    value: function hide() {
      var push = arguments[0] === undefined ? true : arguments[0];

      if (_get(Object.getPrototypeOf(AjaxOverlay.prototype), 'hide', this).call(this) && !!push) {
        this.router.pushRoot();
      }
    }
  }, {
    key: 'handleRoutePop',

    // Handle browser popstate event
    // @param {Object} state  The state object associated with the popstate event
    value: function handleRoutePop(state) {
      if (!!state.title && state.title.length > 0) {
        document.title = state.title;
      }

      if (!!state && !!state.shown) {
        this.show(state.url, state.title, false);
      } else {
        this.hide(false);
      }
    }
  }]);

  return AjaxOverlay;
})(_overlay2['default']);

exports['default'] = AjaxOverlay;
module.exports = exports['default'];

},{"./overlay":3,"./router":4}],3:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var Overlay = (function () {

  // @constructor
  // @param {HTMLElement} el
  // @param {Object}      options

  function Overlay() {
    var el = arguments[0] === undefined ? document.getElementsByClassName('js-overlay')[0] : arguments[0];
    var options = arguments[1] === undefined ? { selectors: {}, states: {} } : arguments[1];

    _classCallCheck(this, Overlay);

    this.el = el;

    this.options = {
      selectors: {
        content: '.js-overlay-content',
        hide: '.js-overlay-hide',
        show: '.js-overlay-show',
        toggle: '.js-overlay-toggle'
      },
      states: {
        hidden: 'is-hidden',
        shown: 'is-shown'
      }
    };

    // Merge provided selectors
    for (var selector in options.selectors) {
      if (options.selectors.hasOwnProperty(selector)) {
        this.options.selectors[selector] = options.selectors[selector];
      }
    }

    // Merge provided states
    for (var state in options.states) {
      if (options.states.hasOwnProperty(state)) {
        this.options.states[state] = options.states[state];
      }
    }

    this.init().bind();
  }

  _createClass(Overlay, [{
    key: 'init',

    // Initial setup
    value: function init() {
      return this;
    }
  }, {
    key: 'bind',

    // Bind event handlers to certain DOM elements, specified through
    value: function bind() {
      var _this = this;

      document.body.addEventListener('keyup', function (e) {
        if (e.keyCode === 27) {
          _this.hide();
        }
      });

      this.bindDelegate('click', this.options.selectors.show, function (e) {
        e.preventDefault();
        _this.show();
      });

      this.bindDelegate('click', this.options.selectors.hide, function (e) {
        e.preventDefault();
        _this.hide();
      });

      this.bindDelegate('click', this.options.selectors.toggle, function (e) {
        e.preventDefault();
        _this.toggle();
      });

      return this;
    }
  }, {
    key: 'bindDelegate',

    // Helper method to support clicks on elements that don't exist yet
    // @param {String}    evt       Event to listen for
    // @param {String}    selector  Element selector that triggers the event
    // @param {Function}  handler   Callback method
    value: function bindDelegate(evt, selector, handler) {
      var matches = document.body.matches || document.body.webkitMatchesSelector || document.body.msMatchesSelector;

      // Listen to all events of type evt
      document.body.addEventListener(evt, function (e) {
        // If e.target matches the specified selector or is a child element
        if (!!matches.call(e.target, '' + selector + ', ' + selector + ' *')) {
          // Find the actual element that matches the selector
          var matchingElement = e.target;
          while (!matches.call(matchingElement, selector)) {
            matchingElement = matchingElement.parentNode;
          }

          // Execute handler, pass on the Event object
          // and have the matching element as the second argument
          handler(e, matchingElement);
        }
      }, false);
    }
  }, {
    key: 'show',

    // Show the overlay
    value: function show() {
      if (this.el.className.indexOf(this.options.states.shown) !== -1) {
        return false;
      }

      var regex = new RegExp('\\s*' + this.options.states.hidden + '\\s*', 'gi'),
          // Template string not working because of reasons
      evt = new CustomEvent('overlay:show', { bubbles: true, cancelable: true }); // @fixme - IE9 support

      this.el.className = this.el.className.replace(regex, '') + ' ' + this.options.states.shown;

      document.body.dispatchEvent(evt);

      return true;
    }
  }, {
    key: 'hide',

    // Hide the overlay
    value: function hide() {
      if (this.el.className.indexOf(this.options.states.hidden) !== -1) {
        return false;
      }

      var regex = new RegExp('\\s*' + this.options.states.shown + '\\s*', 'gi'),
          // Template string not working because of reasons
      evt = new CustomEvent('overlay:hide', { bubbles: true, cancelable: true }); // @fixme - IE9 support

      this.el.className = this.el.className.replace(regex, '') + ' ' + this.options.states.hidden;

      document.body.dispatchEvent(evt);

      return true;
    }
  }, {
    key: 'toggle',

    // Toggle the overlay
    value: function toggle() {
      if (this.el.className.indexOf(this.options.states.shown) === -1) {
        this.show();
      } else {
        this.hide();
      }
    }
  }, {
    key: 'render',

    // Render HTML inside the content selector
    value: function render(html) {
      this.el.querySelector(this.options.selectors.content).innerHTML = html;
    }
  }]);

  return Overlay;
})();

exports['default'] = Overlay;
module.exports = exports['default'];

},{}],4:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var Router = (function () {
  function Router(handlers) {
    _classCallCheck(this, Router);

    this.isSupported = !!history && !!history.pushState;
    this.root = {};

    this.handlers = {
      pop: handlers.pop || function () {}
    };

    this.init().bind();
  }

  _createClass(Router, [{
    key: 'init',
    value: function init() {
      this.root = {
        url: window.location.pathname,
        title: document.title
      };

      this.replace(this.root.url, { title: this.root.title });
      return this;
    }
  }, {
    key: 'bind',
    value: function bind() {
      var _this = this;

      window.addEventListener('popstate', function (e) {
        console.log('ja dus');
        _this.handlers.pop(e.state);
      });

      return this;
    }
  }, {
    key: 'push',
    value: function push(url) {
      var state = arguments[1] === undefined ? {} : arguments[1];

      if (this.isSupported) {
        state.url = url;
        state.title = state.title || '';

        if (state.title.length > 0) {
          document.title = state.title;
        }

        window.history.pushState(state, state.title, url);
      }
    }
  }, {
    key: 'replace',
    value: function replace(url) {
      var state = arguments[1] === undefined ? {} : arguments[1];

      if (this.isSupported) {
        state.url = url;
        state.title = state.title || '';

        if (state.title.length > 0) {
          document.title = state.title;
        }

        window.history.replaceState(state, state.title, url);
      }
    }
  }, {
    key: 'pushRoot',
    value: function pushRoot() {
      this.push(this.root.url, { url: this.root.url, title: this.root.title });
    }
  }, {
    key: 'setRoot',
    value: function setRoot(url) {
      var title = arguments[1] === undefined ? '' : arguments[1];

      this.root.url = url;
      this.root.title = title;
    }
  }]);

  return Router;
})();

exports['default'] = Router;
module.exports = exports['default'];

},{}]},{},[1])


//# sourceMappingURL=build.js.map